src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,122 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWING_COEFFICIENT = 1.5;
+    private static final int NON_EXISTENT_INDEX = -1;
+    private Object[] values;
+    private int lengthCounter;
+    private int currentCapacity;
+
+    public ArrayList() {
+        values = (T[]) new Object[DEFAULT_CAPACITY];
+        currentCapacity = DEFAULT_CAPACITY;
+    }
+
     @Override
     public void add(T value) {
-
+        grow(currentCapacity);
+        values[lengthCounter] = value;
+        lengthCounter++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > lengthCounter || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " out of bounds");
+        }
+        grow(currentCapacity);
+        System.arraycopy(values, index, values, index + 1, lengthCounter - index);
+        values[index] = value;
+        lengthCounter++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) values[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        final T value = (T) values[index];
+        int newLengthCounter = lengthCounter - 1;
+        if (newLengthCounter >= index) {
+            System.arraycopy(values, index + 1, values, index, newLengthCounter - index);
+        }
+        values[newLengthCounter] = null;
+        lengthCounter--;
+        return value;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = NON_EXISTENT_INDEX;
+        if (element == null) {
+            lengthCounter--;
+            return null;
+        }
```

Comment:
An ArrayList elements can be an object reference or the value null, therefore is no need for this condition